### PR TITLE
mock: document public API in collector module

### PR DESCRIPTION
**Pull request by hds**
## Motivation

There has been interest around publishing `tracing-mock` to crates.io
for some time. In order to make this possible, documentation and some
code clean up is needed.

## Solution

This change adds documentation to the collector module itself and to all the
public APIs in the module. This includes doctests on all the methods
that serve as examples.

Additionally the implementation for the `Expect` struct has been moved
into the module with the definition, this was missed in #2369.

Refs: #539

### CODE REVIEW

File: tracing-mock/src/collector.rs [hunk 1]
```diff
@@ -311,7 +313,7 @@ where
     ///
     /// [`Collect::drop_span`]: fn@tracing::Collect::drop_span
     /// [`close_span`]: fn@Self::close_span
-    #[deprecated(since = "0.1.0", note = "use `MockCollector::close_span` instead")]
```
Comment (hawkw) [inline tracing-mock/src/collector.rs]: why was the deprecated attribute on this function removed?

Comment (hds) [reply tracing-mock/src/collector.rs]: Adding the deprecated attribute meant adding `#[allow_deprecated]` on every single usage of it throughout the tracing tests. I felt it better to leave the deprecated attribute off until I can add a replacement. Then the current usage of `drop_span` can be replaced and this method can be marked as deprecated.

File: tracing-mock/src/collector.rs [hunk 2]
```diff
@@ -37,15 +69,23 @@ struct Running<F: Fn(&Metadata<'_>) -> bool> {
...
     expected: VecDeque<Expect>,
     max_level: Option<LevelFilter>,
     filter: F,
     name: String,
 }
 
+/// A handle which is used to invoke validation of expectations.
 pub struct MockHandle(Arc<Mutex<VecDeque<Expect>>>, String);
 
+/// Create a new [`MockCollector`].
```
[the remark this answers is not in the document]
Comment (hds) [reply tracing-mock/src/collector.rs:203]: I've gone with both. Duplicating one of the more complete examples and also linking the module level docs.

File: tracing-mock/src/collector.rs [hunk 3]
```diff
@@ -75,50 +115,268 @@ where
...
+    /// let (collector, handle) = collector::mock()
+    ///     .named("subscriber-1")
+    ///     .event(expect::event())
+    ///     .run_with_handle();
+    ///
+    /// with_default(collector, || {
+    ///     tracing::info!("a");
+    /// });
+    ///
+    /// handle.assert_finished();
```
[the remark this answers is not in the document]
Comment (hds) [reply tracing-mock/src/collector.rs]: Done, I replaced the existing example with one that uses two collectors.

File: tracing-mock/src/collector.rs [hunk 4]
```diff
@@ -67,58 +270,502 @@ where
...
+    ///
+    /// let (collector, handle) = collector::mock()
+    ///     .event(expect::event())
+    ///     .run_with_handle();
+    ///
+    /// tracing::collect::with_default(collector, || {
+    ///     tracing::info!("a");
+    /// });
+    ///
+    /// # Examples
```
Comment (hds) [inline tracing-mock/src/collector.rs:354]:
```suggestion
```

**Comment by davidbarsky** [comment]
I'm not really gonna be able to review the documentation for this; I trust that Eliza has gotten most of this.

**Review by hawkw — APPROVED** [review]

**Pull request merged**